Blocks: Gathering Tweetstorms

pento opened the pull request:
A way to import tweetstorms directly into the block editor.

![tweetstorm](https://user-images.githubusercontent.com/3629020/85486903-ab282e00-b61f-11ea-8a12-05a044ebb275.gif)

#### Changes proposed in this Pull Request:
* Extends the core Twitter embed block, to allow importing tweetstorms.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* New feature. p1HpG7-9mC-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

##### Setup
* Create a [Jurassic Ninja sandbox](https://jurassic.ninja/create/?jetpack-beta&branch=add/tweetstorm-gatherer) against this PR.
* Go through the Jetpack connection process, select a free plan.

##### Testing
* Paste a link to the first tweet in a tweetstorm into a new post.
* Wait for the embed to show.
* Click on the 'Unroll' option in the toolbar.
* Try importing a bunch of different tweetstorms, see what breaks.

#### Proposed changelog entry for your changes:
* Twitter Embeds: Add support for importing tweetstorm content directly into the post.

jeherve commented:
Do you think you could post on +jetpackp2 about this? PCYsg-iiu-p2/#2-loop-in-the-jetpack-division has some more details on this. 

Thank you!

aaronrobertshaw commented:
@pento Any chance you could expand upon the testing instructions for me?

aaronrobertshaw commented:
This all tests well for me 🙂

aaronrobertshaw commented on _inc/lib/class-jetpack-tweetstorm-helper.php:
```diff
@@ -0,0 +1,71 @@
+			return new WP_Error(
+				'unavailable_site_id',
+				__( 'Sorry, something is wrong with your Jetpack connection.', 'jetpack' ),
+				403
```
This might be a silly question, but is `403` the best status here?

pento replied:
I copy/pasted this function from `Jetpack_Instagram_Gallery_Helper`, which probably got it from `WPCOM_REST_API_V2_Endpoint_Mailchimp`, I didn't really think about the error code it's using. 😉

The site ID being unavailable could be caused by a number of things, some of them are appropriate for a `403` status (as the message suggests, if  the Jetpack connection needs fixing, `403` is correct). Alternatively, `502` or a generic `500` response may be more accurate for whatever has caused this error.

I don't think this method the right place to be debugging why there's no valid site id, even though the site has been connected to WPCOM (as is evident by the block loading), so `403` is a reasonable approximation of what could be causing it.

aaronrobertshaw replied:
I could see cases where it fit and some where it might not have so well. I agree that it is likely the best approximation as you say.

Thanks for the explanation 🙂

pento commented:
Thanks for the review, and updating the description, @glendaviesnz!

I'm happy with where this is at, let's run with it.

jeherve commented on extensions/blocks/gathering-tweetstorms/editor.js:
```diff
@@ -0,0 +1,66 @@
+ * @returns {object} The blockSettings, with our extra functionality inserted.
+ */
+const addTweetstormToTweets = blockSettings => {
+	// Bail if this is not the Twitter embed block, or if the hook has been triggered by a deprecation.
```
Since the block relies on a connection to the WordPress.com API, maybe we should avoid displaying it on sites that are not connected to WordPress.com yet, such as sites using Jetpack's development mode? This would avoid messages like this one on sites that cannot be connected to WordPress.com:
![image](https://user-images.githubusercontent.com/426388/85782590-f5e09d80-b726-11ea-9a16-21671bf4e459.png)
I think you should be able to get it from the initial state?

`get( getJetpackData(), [ 'jetpack', 'is_active' ], false );`

pento replied:
I didn't think blocks loaded when Jetpack isn't connected? 

I've added a clearer message for dev mode in c25778207.

jeherve replied:
Thank you!

> I didn't think blocks loaded when Jetpack isn't connected?

If Development mode is active, we do load blocks. This allows one to test blocks that do not require a connection to WordPress.com from their local environment. That's why we end up having to check for a connection for blocks like the Instagram one:
https://github.com/Automattic/jetpack/blob/5f098a65b1621103d2e9c288009f528425533ec9/extensions/blocks/instagram-gallery/instagram-gallery.php#L25

pento commented on extensions/blocks/gathering-tweetstorms/editor.js:
```diff
@@ -0,0 +1,66 @@
+ * Internal dependencies
+ */
+import useGatherTweetstorm from './use-gather-tweetstorm';
+import { withNotices, ToolbarButton, ToolbarGroup } from '@wordpress/components';
```
Added a fallback for this in 83b9978ae.

pento commented:
Thanks for the review, @jeherve!

> When I start from a tweet that's not the first in the thread, would it be possible to fetch the earlier tweets in that thread? Right now it only seems to pick up tweets that follow the one I pasted.

Good catch, I've implemented this in D45614-code.

> Should we try to sideload image embeds instead of hotlinking from Twitter? That would be a nice to really save Tweetstorms. Maybe we could do that only when we can find that the thread is yours?

I think this is out of scope of this PR. There would be a substantial amount of work involved in sideloading images reliably, it would only become worse with videos (see: everything to do with importers).

> When not using the Gutenberg plugin, styles seem a bit off:

Fixed in 83b9978.

> I expected the YouTube videos to turn into embed blocks but that didn't quite work.

Also fixed in D45614-code.

pento commented on extensions/blocks/gathering-tweetstorms/use-gather-tweetstorm.js:
```diff
@@ -0,0 +1,126 @@
+							case 'paragraph':
+								return createBlock( 'core/paragraph', { content: block.content } );
+							case 'gallery':
+								return createBlock( 'core/gallery', { images: block.images } );
```
This isn't possible without sideloading, since tiled gallery requires images to be in the media library. I agree that it would be a nice upgrade, though.

jeherve commented:
r209740-wpcom (noting that the block is still only in Beta on WordPress.com. We can switch it to Production on Jetpack release day. Until then, folks with access to Beta blocks on WordPress.com will be able to test the new feature there as well).

pento commented:
🎉 Thanks, @jeherve!
